### ISSUE TITLE
HNS endpoints leaked after a pod startup failure

### ISSUE THREAD

**Issue by fasaxc** [kubernetes/kubernetes#68511]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
/kind bug
> /kind feature

**What happened**:

Testing k8s 1.11.2 on Windows with flannel networking (not sure if it affects other plugins).  If a pod fails to start (for example because it's a linux pod that was mistakenly scheduled to a Windows node), I see the HNS endpoints that were created for it get leaked.

**What you expected to happen**:

All resources of failed pod should be cleaned up.

**How to reproduce it (as minimally and precisely as possible)**:

- Start a cluster with at least one Windows node.  I used the flannel network plugin in host-gw mode.  
- Try to run a linux pod on windows: 
```
kubectl run -it --rm bb-a --image=busybox --overrides='{"spec": {"template":{ "spec":{"nodeSelector":{"beta.kubernetes.io/os": "windows"}}}}}'
```
- On the windows node, watch the number of endpoints increase: 
```
PS > Get-HNSEndpoint | measure
```

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): 
```
Client Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.2", GitCommit:"bb9ffb1654d4a729bb4cec18ff088eacc153c239", GitTreeState:"clean", BuildDate:"2018-08-07T23:17:28Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.2", GitCommit:"bb9ffb1654d4a729bb4cec18ff088eacc153c239", GitTreeState:"clean", BuildDate:"2018-08-07T23:08:19Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration: AWS
- OS (e.g. from /etc/os-release): Windows 1803 with hotfixes:
```
Source        Description      HotFixID      InstalledBy          InstalledOn
------        -----------      --------      -----------          -----------
EC2AMAZ-5E... Update           KB4343669     NT AUTHORITY\SYSTEM  7/11/2018 12:00:00 AM
EC2AMAZ-5E... Security Update  KB4343909     NT AUTHORITY\SYSTEM  8/15/2018 12:00:00 AM
```
- Kernel (e.g. `uname -a`):
- Install tools: kubeadm for Linux, manual on Windows
- Others:

**Comment by fasaxc** [comment]
@kubernetes/sig-windows-bugs

**Comment by fasaxc** [comment]
Log from the kubelet: https://gist.github.com/fasaxc/15c13a3047755657232e6acfa2b2e366

**Comment by PatrickLang** [comment]
@fasaxc - what state was the pod in when you captured the logs? Had it been deleted, or was it still in CrashLoopBackoff?

**Comment by daschott** [comment]
I assume when you stop the deployment / scale it down to 0, the endpoints still persist as well?

I tried to repro this on RS5 using Kubernetes built from source and # endpoints stayed the same.

**Comment by fasaxc** [comment]
@daschott Yes, deleting the deployment leaves the endpoints in place on my system.

**Comment by daschott** [comment]
I am able to repro this on 1.11.2 now, but not on v1.13.0-alpha.0.1194+8d1127cb41c939. I suspect this might be fixed with https://github.com/kubernetes/kubernetes/pull/66587 -- @fasaxc, can you confirm?

**Comment by fasaxc** [comment]
@daschott Confirmed, the issue seems fixed if I test with recent master: `v1.13.0-alpha.0.1226+426ef9d349bb3a`, in my case.

**Comment by dineshgovindasamy** [comment]
@fasaxc can we please close this issue?
cc @PatrickLang @daschott
